receiver/googlecloudpubsubreceiver/internal/handler.go

```diff
@@ -0,0 +1,237 @@
+	pubsubpb "google.golang.org/genproto/googleapis/pubsub/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type StreamHandler struct {
+	stream      pubsubpb.Subscriber_StreamingPullClient
+	pushMessage func(ctx context.Context, message *pubsubpb.ReceivedMessage) error
+	acks        []string
+	mutex       sync.Mutex
+	client      *pubsub.SubscriberClient
+
+	clientID     string
+	subscription string
+
+	cancel            context.CancelFunc
+	wg                sync.WaitGroup
+	receiverWaitGroup sync.WaitGroup
+	logger            *zap.Logger
+	duration          time.Duration
+
+	isRunning atomic.Bool
+}
+
+func (handler *StreamHandler) ack(ackID string) {
+	handler.mutex.Lock()
+	defer handler.mutex.Unlock()
+	handler.acks = append(handler.acks, ackID)
+}
+
+func NewHandler(
+	ctx context.Context,
+	logger *zap.Logger,
+	client *pubsub.SubscriberClient,
+	clientID string,
+	subscription string,
+	callback func(ctx context.Context, message *pubsubpb.ReceivedMessage) error) (*StreamHandler, error) {
+
+	handler := StreamHandler{
+		logger:       logger,
+		client:       client,
+		clientID:     clientID,
+		subscription: subscription,
+		pushMessage:  callback,
+		acks:         make([]string, 0),
+		duration:     10000 * time.Millisecond,
+	}
+	return &handler, handler.initStream(ctx)
+}
+
+func (handler *StreamHandler) initStream(ctx context.Context) error {
+	var err error
+	// Create a stream, but with the receivers context as we don't want to cancel and ongoing operation
+	handler.stream, err = handler.client.StreamingPull(ctx)
+	if err != nil {
+		return err
+	}
+
+	request := pubsubpb.StreamingPullRequest{
+		Subscription:             handler.subscription,
+		StreamAckDeadlineSeconds: 60,
+		ClientId:                 handler.clientID,
+	}
+	if err := handler.stream.Send(&request); err != nil {
+		_ = handler.stream.CloseSend()
+		return err
+	}
+	return nil
+}
+
+func (handler *StreamHandler) RecoverableStream(ctx context.Context) {
+	handler.receiverWaitGroup.Add(1)
+	handler.isRunning.Swap(true)
+	var handlerCtx context.Context
+	handlerCtx, handler.cancel = context.WithCancel(ctx)
+	go handler.recoverableStream(handlerCtx)
+}
+
+func (handler *StreamHandler) recoverableStream(ctx context.Context) {
+	for handler.isRunning.Load() {
+		// Create a new cancelable context for the handler, so we can recover the stream
+		var loopCtx context.Context
+		loopCtx, cancel := context.WithCancel(ctx)
+
+		handler.logger.Info("Starting Streaming Pull")
+		handler.wg.Add(2)
+		go handler.requestStream(loopCtx, cancel)
+		go handler.responseStream(loopCtx, cancel)
+
+		select {
+		case <-loopCtx.Done():
+			handler.wg.Wait()
+		case <-ctx.Done():
+			cancel()
+			handler.wg.Wait()
+		}
+		if handler.isRunning.Load() {
+			err := handler.initStream(ctx)
+			if err != nil {
+				handler.logger.Error("Failed to recovery stream.")
+			}
+		}
+		handler.logger.Warn("End of recovery loop, restarting.")
+		time.Sleep(250)
```

Comment:
nit: can we make this a descriptive constant?  e.g. `streamRecoveryBackoffPeriod` or something appropriate